### PR TITLE
Feature/unique ptr in shm copy ir 1115

### PR DESCRIPTION
**Pull request by Alexey-N-Chernyshov**
### Description of the Change
This PR introduces Cloneable interface and clone() function that returns unique_ptr described here (https://github.com/CppCodeReviewers/Covariant-Return-Types-and-Smart-Pointers) to replace shared_model *copy() method that returns raw pointer.

### Benefits
Safety with unique_ptr.

### Possible Drawbacks 
no

### Usage Examples or Tests *[optional]*
```
 // Inherit from clonable with CRTP
 struct Base : object::cloneable<Base>
 {
   // ... other methods
 };

 // override clone() where it is needed
 struct Derived : Base
 {
   // ... other methods
 protected:
   Derived* clone() const override
   {
     return new Derived(*this);
   }
 };

 // and clone with function clone(obj) or clone(obj *)
 Derived derived;
 auto c1 = clone(derived);
 auto c2 = clone(&derived);
```

### Alternate Designs *[optional]*
1. Just override copy() - this approach will hide base.
2. copy() returns unique_ptr<Base> - static_cast needed every time.
3. The same as implemented, but use method instead of function - less general and need to hide base method in every class in hierarchy of shared_model.

### CODE REVIEW

File: example/config.sample [hunk 1]
```diff
@@ -8,4 +8,3 @@
   "vote_delay" : 5000,
   "load_delay" : 5000
 }
-
```
Comment (l4l) [inline example/config.sample]: Revert that plz

File: libs/common/cloneable.hpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright Soramitsu Co., Ltd. 2017 All Rights Reserved.
```
Comment (l4l) [inline libs/common/cloneable.hpp]: 2018

File: libs/common/cloneable.hpp [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+/**
+ * Helper function to copy from pointer to Cloneable.
+ * @tparam T - derived from Cloneable
+ * @param object - object to clone
+ * @return clone of object
+ */
+template <typename T>
+auto clone(T *object) -> decltype(clone(*object)) {
```
Comment (l4l) [inline libs/common/cloneable.hpp]: It doesn't work w/o decltype? Afair c++14 allow omit it

File: shared_model/interfaces/base/model_primitive.hpp [hunk 4]
```diff
@@ -29,7 +30,7 @@ namespace shared_model {
      * @tparam Model - your new style model
      */
     template <typename Model>
-    class ModelPrimitive {
+    class ModelPrimitive : public Cloneable<ModelPrimitive<Model>> {
```
Comment (l4l) [inline shared_model/interfaces/base/model_primitive.hpp]: What is the point in moving away cloning function from ModelPrimitive?

File: test/integration/client_test.cpp [hunk 5]
```diff
@@ -218,9 +218,8 @@ TEST_F(ClientServerTest, SendQueryWhenValid) {
   auto account_admin = iroha::model::Account();
   account_admin.account_id = "admin@test";
 
-  auto account_test = std::shared_ptr<shared_model::interface::Account>(
-      shared_model::proto::AccountBuilder().accountId("test@test").build().copy()
-  );
+  std::shared_ptr<shared_model::interface::Account> account_test = clone(
```
Comment (l4l) [inline test/integration/client_test.cpp]: move?

File: test/module/irohad/model/command_validate_execute_test.cpp [hunk 6]
```diff
@@ -1514,7 +1506,7 @@ TEST_F(TransferAssetTest, InvalidWhenAmountOverflow) {
                               transfer_asset->asset_id))
       .WillOnce(Return(dst_wallet));
 
-  // More than account balance
+  // More than account balance)
```
Comment (l4l) [inline test/module/irohad/model/command_validate_execute_test.cpp]: .

File: shared_model/interfaces/base/model_primitive.hpp (hunk 4 above)
Comment (Alexey-N-Chernyshov) [reply shared_model/interfaces/base/model_primitive.hpp]: To use it somewhere else besides shared_model

Comment (l4l) [reply shared_model/interfaces/base/model_primitive.hpp]: Not sure that it is needed

File: test/integration/client_test.cpp (hunk 5 above)
Comment (Alexey-N-Chernyshov) [reply test/integration/client_test.cpp]: moving a temporary object prevents copy elision

File: libs/common/cloneable.hpp [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+ public:
+  using base_type = T;
+
+  virtual ~Cloneable() = default;
+
+ protected:
+  /**
+   * Polymorphic clone constructor.
+   * Method guarantees deep-copy.
+   * @return pointer to clone object
```
Comment (l4l) [inline libs/common/cloneable.hpp]: clone**d**

**Pull request merged**